### PR TITLE
Reason mode cleanup and refactoring

### PR DESCRIPTION
**Pull request by fredyr**
- Removed the remains of the rust mode origins
- Moved the utop integration into reason-interaction.el
- Indentation support moved into reason-indent
- First JSX syntax support
- Boolean constanst true/false highlighted as constants instead of keywords
- Added missing keywords: external/mutable/where/open

### CODE REVIEW

File: reason-mode.el [hunk 1]
```diff
@@ -12,123 +12,17 @@
 ;; This file is distributed under the terms of both the MIT license and the
 ;; Apache License (version 2.0).
 
-;;; Commentary:
```
Comment (arichiardi) [inline reason-mode.el]: Oh I spent a full 10 minutes to write this and now it goes away? Can we restore it?

File: reason-interaction.el [hunk 2]
```diff
@@ -0,0 +1,208 @@
+
```
Comment (arichiardi) [inline reason-interaction.el:3]: Probably we need to add the header like in `reason-mode.el` in all files (or my emacs complains):

```
;;; reason-mode.el --- A major emacs mode for editing Reason (based on rust-mode) -*-lexical-binding: t-*-
  ;; Portions Copyright (c) 2015-present, Facebook, Inc. All rights reserved.
```

File: reason-indent.el [hunk 3]
```diff
@@ -0,0 +1,259 @@
+(defcustom reason-indent-offset 2
```
Comment (arichiardi) [inline reason-indent.el:4]: Can we add the header? See other comment.

File: reason-mode.el [hunk 4]
```diff
@@ -1354,15 +174,17 @@ This is written mainly to be used as `end-of-defun-function' for Reason."
...
 
-
 (defun reason-mode-reload ()
   (interactive)
   (unload-feature 'reason-mode)
+  (unload-feature 'reason-indent)
   (require 'reason-mode)
   (reason-mode))
 
 (require 'refmt)
```
Comment (arichiardi) [inline reason-mode.el]: Let's keep the `require` on the top. Not that it matters, but I feel it is better to see what it is loaded up there.

File: reason-indent.el [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+                      nil 'move (or arg 1)))
+
+(defun reason-end-of-defun ()
+  "Move forward to the next end of defun.
+
+With argument, do it that many times.
+Negative argument -N means move back to Nth preceding end of defun.
+
+Assume that this is called after beginning-of-defun. So point is
+at the beginning of the defun body.
```
Comment (arichiardi) [inline reason-indent.el:65]: Shall we error out or handle the case this is not the case?

File: reason-mode.el [hunk 6]
```diff
@@ -1354,15 +174,17 @@ This is written mainly to be used as `end-of-defun-function' for Reason."
...
-
+(require 'reason-interaction)
 (provide 'reason-mode)
 
 ;;; reason-mode.el ends here
+
+;;; TODO: keymaps, e.g. (define-key map "\C-c\C-a" 'reason-mode-find-alternate-file)
```
Comment (arichiardi) [inline reason-mode.el:207]: definitely

**Review by arichiardi — CHANGES_REQUESTED** [review]

File: reason-mode.el (hunk 1 above)
Comment (fredyr) [reply reason-mode.el]: My bad.

File: reason-indent.el (hunk 5 above)
Comment (fredyr) [reply reason-indent.el:65]: reason-indent is just the indent parts pulled out of reason-mode.  these functions are as in the current version. sounds good to improve this as things progress, but for now i wouldn't now how.

Comment (arichiardi) [reply reason-indent.el:65]: Yeah for sure, I thought it was added brand new :wink:

**Comment by fredyr** [comment]
I've updated as per your requests. Let me know if there's anything else.

**Comment by arichiardi** [comment]
Awesome thanks, will try it out and merge when ready, good job!

**Review by arichiardi — APPROVED** [review]

**Pull request merged**